Upwards facing liquid mesh culling

Obverser opened the pull request:
# Description

Addresses an issue where liquids would not render the top face of their mesh when a block is directly above.
Fixes #47 

- [x] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation comments
- [x] My changes generate no new warnings

stackotter commented on Sources/Core/Sources/Render/Mesh/FluidMeshBuilder.swift:
```diff
@@ -106,9 +106,10 @@ struct FluidMeshBuilder { // TODO: Make fluid meshes look more like they do in v
-    // Iterate through all visible faces
-    for direction in Direction.allDirections where !cullingNeighbours.contains(direction) {
+    // Iterate through all visible faces, filtering out the top face if necessary
+    for direction in Direction.allDirections where !cullingNeighbours.filter({d in  d != .up || cullTop }).contains(direction) {
```
The filter will get run every iteration and is an inefficient operation to use on a set. Instead, make a mutable copy of the set and remove `.up` if `!cullTop` (before the for loop). In the future I'll make this even more efficient by using a bitset (OptionSet) instead of a `Set<Direction>`, but that's not necessary for now.

Obverser replied:
Actually, is it possible to just hook the conditional into the `case .up:` statement? Something where we just run the case if cullTop is true?

Obverser replied:
Should be more efficient now. I'm going to look into the linux branch now.

stackotter reviewed (changes requested):
Looking good so far. Just have one requested change to avoid hurting the performance of the modified fluid mesh builder method.

Also, if it's not too much of a pain, would you be able to rebase onto the `linux` branch? I've moved all of the rendering code into a separate target so it'll cause a merge conflict, but you'll really only have to copy the changes from the old location to the new location.

Obverser commented:
GitHub probably thinks I'm a psychopath now with the two forced updates...

Rebased for the linux branch and changed the pull request! I'm pretty sure it works, but I don't have a linux machine immediately accessible to test on.

ninjadev64 commented:
> I'm pretty sure it works, but I don't have a linux machine immediately accessible to test on.

As long as it's working on macOS, because there isn't a renderer for Linux yet (only a chat client).

The reason for the magnitude of rendering changes on the Linux branch is due to the split of DeltaCore into separate Core and Renderer targets, so that the Metal renderer wouldn't prohibit DeltaCore from building on Linux.

stackotter commented on Sources/Core/Renderer/Mesh/FluidMeshBuilder.swift:
```diff
@@ -107,9 +107,11 @@ struct FluidMeshBuilder { // TODO: Make fluid meshes look more like they do in v
     tint: Vec3f
   ) {
     let basePosition = position.relativeToChunkSection.floatVector + Vec3f(0.5, 0, 0.5)
+    var cullingFluid = cullingNeighbours
```
The variable name `cullingFluid` doesn't really make sense for a Set to me. Maybe just do `var cullingNeighbours = cullingNeighbours` instead

stackotter commented on Sources/Core/Renderer/Mesh/FluidMeshBuilder.swift:
```diff
@@ -107,9 +107,11 @@ struct FluidMeshBuilder { // TODO: Make fluid meshes look more like they do in v
   ) {
     let basePosition = position.relativeToChunkSection.floatVector + Vec3f(0.5, 0, 0.5)
+    var cullingFluid = cullingNeighbours
+    if neighbouringBlocks[.up]?.fluidId != fluid.id { cullingFluid.remove(.up) }
```
Avoid single line if statements, I find it worth while (especially while skim-reading) to use the extra two lines.

Obverser replied:
That's my bad, I rushed and forgot that you had the whole conventions document attached. I'll do a formatting commit in just a moment

stackotter reviewed (changes requested):
Just some style/formatting changes this time

Obverser commented:
Should've checked back on the formatting guide. Here's a better commit

stackotter reviewed (approved):
Looks good to me 👍 thanks for fixing this
